File: Week01/vanilla_implementations/sieve_of_eratosthenes.py
# Sieve of Eratosthenes implementation in Python
#
# Author: Bedir Tapkan
# Date:   04/11/2022

def sieve_of_eratosthenes(n):
    """
    Sieve of Eratosthenes implementation;
    Finding prime numbers from 2 to n (inclusive).
    """
    primes = [True] * (n + 1)
    primes[0] = False
    primes[1] = False
    results = [] # List to store the prime numbers
    for i in range(2, int(n ** 0.5) + 1):
        if primes[i]:
            results.append(i)
            for j in range(i * i, n + 1, i):
                primes[j] = False
    return results + [i for i in range(int(n ** 0.5) + 1, n + 1) if primes[i]]

def main():
    """
    Main function
    """
    n = int(input("n: "))
    primes = sieve_of_eratosthenes(n)
    print(primes)


if __name__ == "__main__":
    main()

File: Week01/vanilla_implementations/SieveOfEratosthenes.cpp
// Sieve of Eratosthenes
// 
// Author: Bedir Tapkan
//
// Desc: Find the prime numbers from 1 - n

#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

void sieve(long n, bool * composite, vector<int> &primes){
    // n -> the max number to check for primes
    // composite -> bool arr to mark composite nums
    // primes -> vector that contains primes from 2 -> n

	// O(n*sqrt(n)) - but much faster in practice
	for (long i = 2; i <= n; ++i){ 	
	    // Check 2->n if they are already marked as composite
        // and mark composites along the way
		if (!composite[i]){
			// Checking if prime
			primes.push_back(i);
			for (long j = i*i; j <= n; j += i){
				composite[j] = true;
			}
		}
	}
}
void printVector(vector<int> vc) {
	for (int i = 0; i < vc.size(); ++i)
		cout << vc[i] << " ";
	cout << endl << endl;
}

int main() {
	long n;
	cin >> n; // number to check for primes up to

	bool composite[n+1];
	memset(composite, false, sizeof composite);
	vector<int> primes;

	sieve(n, composite, primes);
	printVector(primes);
}

